File: examples/hello-world/gabriel-greeting-app-kotlin-compose/app/src/test/kotlin/org/organicprogramming/gabriel/greeting/kotlincompose/controller/GreetingControllerTest.kt
package org.organicprogramming.gabriel.greeting.kotlincompose.controller

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test
import org.organicprogramming.gabriel.greeting.kotlincompose.model.AppPlatformCapabilities
import org.organicprogramming.gabriel.greeting.kotlincompose.support.FakeGreetingHolonConnection
import org.organicprogramming.gabriel.greeting.kotlincompose.support.FakeHolonCatalog
import org.organicprogramming.gabriel.greeting.kotlincompose.support.FakeHolonConnector
import org.organicprogramming.gabriel.greeting.kotlincompose.support.holon
import org.organicprogramming.gabriel.greeting.kotlincompose.support.language

class GreetingControllerTest {
    @Test
    fun initializesWithPreferredHolonAndGreetsEnglish() = runBlocking {
        val swiftConnection = FakeGreetingHolonConnection(
            languages = listOf(
                language("en", "English", "English"),
                language("fr", "French", "Francais"),
            ),
            greetingBuilder = { name, _ -> "Hello $name from Swift" },
        )
        val connector = FakeHolonConnector(
            mapOf(
                "gabriel-greeting-go" to { _ ->
                    FakeGreetingHolonConnection(
                        languages = listOf(language("en", "English", "English")),
                        greetingBuilder = { name, _ -> "Hello $name from Go" },
                    )
                },
                "gabriel-greeting-swift" to { _ -> swiftConnection },
            ),
        )
        val controller = GreetingController(
            catalog = FakeHolonCatalog(listOf(holon("gabriel-greeting-go"), holon("gabriel-greeting-swift"))),
            connector = connector,
            initialTransport = "stdio",
        )

        controller.initialize()

        assertEquals("gabriel-greeting-swift", controller.state.value.selectedHolon?.slug)
        assertEquals("en", controller.state.value.selectedLanguageCode)
        assertEquals("Hello World from Swift", controller.state.value.greeting)
        assertEquals(listOf("gabriel-greeting-swift" to "stdio"), connector.connectCalls)
        assertEquals(listOf("World" to "en"), swiftConnection.sayHelloCalls)
    }

    @Test
    fun changingTheUserNameRefreshesTheGreeting() = runBlocking {
        val connection = FakeGreetingHolonConnection(
            languages = listOf(language("en", "English", "English")),
            greetingBuilder = { name, _ -> "Hello $name" },
        )
        val controller = GreetingController(
            catalog = FakeHolonCatalog(listOf(holon("gabriel-greeting-swift"))),
            connector = FakeHolonConnector(mapOf("gabriel-greeting-swift" to { _ -> connection })),
        )

        controller.initialize()
        controller.setUserName("Alice")

        assertEquals("Hello Alice", controller.state.value.greeting)
        assertEquals("Alice" to "en", connection.sayHelloCalls.last())
    }

    @Test
    fun rejectsUnixTransportWhenPlatformCapabilityDisablesIt() {
        runBlocking {
            val controller = GreetingController(
                catalog = FakeHolonCatalog(listOf(holon("gabriel-greeting-swift"))),
                connector = FakeHolonConnector(
                    mapOf(
                        "gabriel-greeting-swift" to { _ ->
                            FakeGreetingHolonConnection(
                                languages = listOf(language("en", "English", "English")),
                                greetingBuilder = { name, _ -> "Hello $name" },
                            )
                        },
                    ),
                ),
                capabilities = AppPlatformCapabilities(supportsUnixSockets = false),
            )

            assertThrows(IllegalArgumentException::class.java) {
                runBlocking {
                    controller.setTransport("unix")
                }
            }
        }
    }

    @Test
    fun surfacesConnectorFailuresAsErrors() = runBlocking {
        val controller = GreetingController(
            catalog = FakeHolonCatalog(listOf(holon("gabriel-greeting-swift"))),
            connector = FakeHolonConnector(
                mapOf(
                    "gabriel-greeting-swift" to { _ ->
                        FakeGreetingHolonConnection(
                            languages = emptyList(),
                            greetingBuilder = { _, _ -> "ignored" },
                            listLanguagesError = IllegalStateException("boot failure"),
                        )
                    },
                ),
            ),
        )

        controller.initialize()

        assertNull(controller.state.value.connectionError)
        assertTrue(controller.state.value.error?.contains("Failed to load languages") == true)
        assertTrue(controller.state.value.isRunning)
    }

    @Test
    fun transportChangeInvalidatesInflightStart() = runBlocking {
        coroutineScope {
        val stdioDeferred = CompletableDeferred<FakeGreetingHolonConnection>()
        val tcpConnection = FakeGreetingHolonConnection(
            languages = listOf(language("en", "English", "English")),
            greetingBuilder = { name, _ -> "Hello $name" },
        )
        val connectCalls = mutableListOf<Pair<String, String>>()
        val controller = GreetingController(
            catalog = FakeHolonCatalog(listOf(holon("gabriel-greeting-swift"))),
            connector = object : org.organicprogramming.gabriel.greeting.kotlincompose.runtime.HolonConnector {
                override suspend fun connect(
                    holon: org.organicprogramming.gabriel.greeting.kotlincompose.model.GabrielHolonIdentity,
                    transport: String,
                ): org.organicprogramming.gabriel.greeting.kotlincompose.runtime.GreetingHolonConnection {
                    connectCalls += holon.slug to transport
                    return when (transport) {
                        "stdio" -> stdioDeferred.await()
                        "tcp" -> tcpConnection
                        else -> error("unexpected transport $transport")
                    }
                }
            },
            initialTransport = "stdio",
        )

        controller.refreshHolons()
        val firstStart = async { controller.ensureStarted() }
        delay(25)
        controller.setTransport("tcp", reload = false)
        controller.ensureStarted()

        assertEquals(
            listOf(
                "gabriel-greeting-swift" to "stdio",
                "gabriel-greeting-swift" to "tcp",
            ),
            connectCalls,
        )
        assertEquals("tcp", controller.state.value.transport)
        assertTrue(controller.state.value.isRunning)

        stdioDeferred.complete(
            FakeGreetingHolonConnection(
                languages = listOf(language("en", "English", "English")),
                greetingBuilder = { name, _ -> "Hello $name from stdio" },
            ),
        )
        firstStart.await()

        assertEquals("tcp", controller.state.value.transport)
        assertTrue(controller.state.value.isRunning)
        }
    }
}

File: examples/hello-world/gabriel-greeting-app-flutter/app/test/greeting_controller_test.dart
import 'dart:async';

import 'package:flutter_test/flutter_test.dart';

import 'package:gabriel_greeting_app_flutter/src/controller/greeting_controller.dart';
import 'package:gabriel_greeting_app_flutter/src/model/app_model.dart';
import 'package:gabriel_greeting_app_flutter/src/runtime/holon_connector.dart';

import 'support/fakes.dart';

void main() {
  group('GreetingController', () {
    test('initializes with preferred holon and greets in English', () async {
      final swiftConnection = FakeGreetingHolonConnection(
        languages: [
          language(code: 'en', name: 'English', native: 'English'),
          language(code: 'fr', name: 'French', native: 'Francais'),
        ],
        greetingBuilder: ({required name, required langCode}) =>
            'Hello $name from Swift',
      );
      final connector = FakeHolonConnector(
        factories: <String, FakeGreetingHolonConnection Function(String)>{
          'gabriel-greeting-go': (_) => FakeGreetingHolonConnection(
            languages: [
              language(code: 'en', name: 'English', native: 'English'),
            ],
            greetingBuilder: ({required name, required langCode}) =>
                'Hello $name from Go',
          ),
          'gabriel-greeting-swift': (_) => swiftConnection,
        },
      );
      final controller = GreetingController(
        catalog: FakeHolonCatalog(<GabrielHolonIdentity>[
          holon('gabriel-greeting-go'),
          holon('gabriel-greeting-swift'),
        ]),
        connector: connector,
        initialTransport: 'stdio',
      );

      await controller.initialize();
      await waitForCoaxUpdate();

      expect(controller.selectedHolon?.slug, 'gabriel-greeting-swift');
      expect(controller.selectedLanguageCode, 'en');
      expect(controller.greeting, 'Hello World from Swift');
      expect(connector.connectCalls, [('gabriel-greeting-swift', 'stdio')]);
      expect(swiftConnection.sayHelloCalls.single, ('World', 'en'));
    });

    test('changing the user name refreshes the greeting', () async {
      final connection = FakeGreetingHolonConnection(
        languages: [language(code: 'en', name: 'English', native: 'English')],
        greetingBuilder: ({required name, required langCode}) => 'Hello $name',
      );
      final controller = GreetingController(
        catalog: FakeHolonCatalog(<GabrielHolonIdentity>[
          holon('gabriel-greeting-swift'),
        ]),
        connector: FakeHolonConnector(
          factories: <String, FakeGreetingHolonConnection Function(String)>{
            'gabriel-greeting-swift': (_) => connection,
          },
        ),
      );

      await controller.initialize();
      await waitForCoaxUpdate();
      await controller.setUserName('Alice');

      expect(controller.greeting, 'Hello Alice');
      expect(connection.sayHelloCalls.last, ('Alice', 'en'));
    });

    test(
      'rejects unix transport when the platform capability disables it',
      () async {
        final controller = GreetingController(
          catalog: FakeHolonCatalog(<GabrielHolonIdentity>[
            holon('gabriel-greeting-swift'),
          ]),
          connector: FakeHolonConnector(
            factories: <String, FakeGreetingHolonConnection Function(String)>{
              'gabriel-greeting-swift': (_) => FakeGreetingHolonConnection(
                languages: [
                  language(code: 'en', name: 'English', native: 'English'),
                ],
                greetingBuilder: ({required name, required langCode}) =>
                    'Hello $name',
              ),
            },
          ),
          capabilities: const AppPlatformCapabilities(
            supportsUnixSockets: false,
          ),
        );

        expect(
          () => controller.setTransport('unix'),
          throwsA(isA<StateError>()),
        );
      },
    );

    test('surfaces connector failures as connection errors', () async {
      final controller = GreetingController(
        catalog: FakeHolonCatalog(<GabrielHolonIdentity>[
          holon('gabriel-greeting-swift'),
        ]),
        connector: FakeHolonConnector(
          factories: <String, FakeGreetingHolonConnection Function(String)>{
            'gabriel-greeting-swift': (_) => FakeGreetingHolonConnection(
              languages: const [],
              greetingBuilder: ({required name, required langCode}) =>
                  'ignored',
              listLanguagesError: StateError('boot failure'),
            ),
          },
        ),
      );

      await controller.initialize();

      expect(controller.connectionError, isNull);
      expect(controller.error, contains('Failed to load languages'));
      expect(controller.isRunning, isTrue);
    });

    test('transport change invalidates an in-flight start', () async {
      final stdioCompleter = Completer<GreetingHolonConnection>();
      final tcpConnection = FakeGreetingHolonConnection(
        languages: [language(code: 'en', name: 'English', native: 'English')],
        greetingBuilder: ({required name, required langCode}) => 'Hello $name',
      );
      final connector = _DeferredHolonConnector(
        onConnect: (transport) {
          if (transport == 'stdio') {
            return stdioCompleter.future;
          }
          if (transport == 'tcp') {
            return Future<GreetingHolonConnection>.value(tcpConnection);
          }
          throw StateError('Unexpected transport $transport');
        },
      );
      final controller = GreetingController(
        catalog: FakeHolonCatalog(<GabrielHolonIdentity>[
          holon('gabriel-greeting-swift'),
        ]),
        connector: connector,
        initialTransport: 'stdio',
      );

      await controller.refreshHolons();
      final firstStart = controller.ensureStarted();
      await Future<void>.delayed(Duration.zero);

      await controller.setTransport('tcp', reload: false);
      await controller.ensureStarted();

      expect(connector.connectCalls, <(String, String)>[
        ('gabriel-greeting-swift', 'stdio'),
        ('gabriel-greeting-swift', 'tcp'),
      ]);
      expect(controller.transport, 'tcp');
      expect(controller.isRunning, isTrue);

      stdioCompleter.complete(
        FakeGreetingHolonConnection(
          languages: [language(code: 'en', name: 'English', native: 'English')],
          greetingBuilder: ({required name, required langCode}) =>
              'Hello $name from stdio',
        ),
      );
      await firstStart;

      expect(controller.transport, 'tcp');
      expect(controller.isRunning, isTrue);
    });

    test(
      'retries non-stdio connection startup once before surfacing failure',
      () async {
        var tcpAttempts = 0;
        final controller = GreetingController(
          catalog: FakeHolonCatalog(<GabrielHolonIdentity>[
            holon('gabriel-greeting-c'),
          ]),
          connector: FakeHolonConnector(
            factories: <String, FakeGreetingHolonConnection Function(String)>{
              'gabriel-greeting-c': (transport) {
                if (transport == 'tcp' && tcpAttempts++ == 0) {
                  throw StateError('temporary tcp startup race');
                }
                return FakeGreetingHolonConnection(
                  languages: [
                    language(code: 'en', name: 'English', native: 'English'),
                  ],
                  greetingBuilder: ({required name, required langCode}) =>
                      'Hello $name',
                );
              },
            },
          ),
          initialTransport: 'tcp',
        );

        await controller.initialize();
        await waitForCoaxUpdate();

        expect(controller.isRunning, isTrue);
        expect(controller.connectionError, isNull);
        expect(controller.greeting, 'Hello World');
        expect(tcpAttempts, 2);
      },
    );
  });
}

class _DeferredHolonConnector implements HolonConnector {
  _DeferredHolonConnector({required this.onConnect});

  final Future<GreetingHolonConnection> Function(String transport) onConnect;
  final List<(String slug, String transport)> connectCalls =
      <(String, String)>[];

  @override
  Future<GreetingHolonConnection> connect(
    GabrielHolonIdentity holon, {
    required String transport,
  }) {
    connectCalls.add((holon.slug, transport));
    return onConnect(transport);
  }
}
